src/overlays/actors/ovl_En_Rat/z_en_rat.c

```diff
@@ -28,123 +44,910 @@ const ActorInit En_Rat_InitVars = {
+        EnRat_Explode(this, play);
+    }
+}
+
+void EnRat_Explode(EnRat* this, PlayState* play) {
+    EnBom* bomb;
+
+    bomb = (EnBom*)Actor_Spawn(&play->actorCtx, play, ACTOR_EN_BOM, this->actor.world.pos.x, this->actor.world.pos.y,
+                               this->actor.world.pos.z, 0, 0, 0, 0);
+    if (bomb != NULL) {
+        bomb->timer = 0;
+    }
+
+    if (EN_RAT_GET_TYPE(&this->actor) == EN_RAT_TYPE_OVERWORLD) {
+        Item_DropCollectibleRandom(play, &this->actor, &this->actor.world.pos, 0x100);
+    }
+
+    this->actor.speedXZ = 0.0f;
+    this->actionFunc = EnRat_PostDetonation;
+}
+
+/**
+ * If the Real Bombchu is an Overworld-type Bombchu, this will set it up to revive.
+ * Otherwise, it will simply kill the Bombchu.
+ */
+void EnRat_PostDetonation(EnRat* this, PlayState* play) {
+    if (EN_RAT_GET_TYPE(&this->actor) == EN_RAT_TYPE_OVERWORLD) {
+        EnRat_SetupRevive(this);
+    } else {
+        Actor_MarkForDeath(&this->actor);
+    }
+}
+
+void EnRat_Update(Actor* thisx, PlayState* play) {
+    s32 pad;
+    EnRat* this = THIS;
+
+    this->shouldRotateOntoSurfaces = false;
+    if (this->stunTimer == 0) {
+        SkelAnime_Update(&this->skelAnime);
+    }
+
+    if (this->collider.base.atFlags & AT_HIT) {
+        this->collider.base.atFlags &= ~AT_HIT;
+        this->collider.base.acFlags &= ~AC_HIT;
+        this->collider.base.ocFlags1 &= ~OC1_HIT;
+        if (this->collider.base.atFlags & AT_BOUNCED) {
+            EnRat_SetupBounced(this);
+        } else {
+            EnRat_Explode(this, play);
+            return;
+        }
+    } else if (this->collider.base.acFlags & AC_HIT) {
+        this->collider.base.acFlags &= ~AC_HIT;
+        if (this->actor.colChkInfo.damageEffect == EN_RAT_DMGEFF_HOOKSHOT) {
+            this->stunTimer = -2;
+        } else if (this->actor.colChkInfo.damageEffect == EN_RAT_DMGEFF_STUN) {
+            Actor_PlaySfxAtPos(&this->actor, NA_SE_EN_COMMON_FREEZE);
+            Actor_SetColorFilter(&this->actor, 0, 120, 0, 40);
+            if (this->actionFunc == EnRat_Bounced) {
+                this->actor.speedXZ = 0.0f;
+                if (this->actor.velocity.y > 0.0f) {
+                    this->actor.velocity.y = 0.0f;
+                }
+            } else {
+                this->stunTimer = 40;
+            }
+        } else {
+            EnRat_Explode(this, play);
+            return;
+        }
+    } else if (((this->collider.base.ocFlags1 & OC1_HIT) && (((this->collider.base.oc->category == ACTORCAT_ENEMY)) ||
+                                                             (this->collider.base.oc->category == ACTORCAT_BOSS) ||
+                                                             (this->collider.base.oc->category == ACTORCAT_PLAYER))) ||
+               ((this->actionFunc == EnRat_ChasePlayer) && (this->timer == 0))) {
+        this->collider.base.ocFlags1 &= ~OC1_HIT;
+        EnRat_Explode(this, play);
+        return;
+    }
```

Comment:
Seems slightly odd to me, actors with this much collision logic usually get their own CheckCollision function instead of update being this big, wonder if this dev was different, new to the project?

Reply:
Yeah this was definitely a weird actor in terms of implementation. I think large chunks of it got copied from EnBomChu without a ton of modification.